bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,132 @@
-            warn!("nodes credentials not set");
+    fn spawn_resolver(&self, cred: DeclaredCredentials) {
+        tokio::spawn(Self::resolve_worker(self.nodes.clone(), cred, self.resolve_sleep_dur_ms));
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, cred: DeclaredCredentials, sleep_dur_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = lookup(hostname).await;
+        let mut cur_sleep_dur_ms = 100;
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
+            tokio::time::sleep(Duration::from_millis(cur_sleep_dur_ms)).await;
+
+            addr = lookup(hostname).await;
+
+            cur_sleep_dur_ms = sleep_dur_ms.min(cur_sleep_dur_ms * 2);
+        }
+
+        let addr = addr.expect("somehow addr is none");
+        if let CredentialsKind::InterNode(nodename) = cred.kind() {
+            let mut nodes = nodes.write().expect("nodes credentials lock");
+            if let Some(creds) = nodes.get_mut(nodename) {
+                creds.creds_mut().replace_addresses(addr);
+                creds.set_resolved();
+            }
+        } else {
+            error!("resolved credentials are not internode");
+        }
+    }
+
+    fn process_auth_result(&self, nodename: &str, unresolved: bool) {
+        let mut nodes = self.nodes.write().expect("nodes credentials lock");
+        if let Some(node) = nodes.get_mut(nodename) {
+            if node.update_resolve_state(unresolved) {
+                node.set_pending();
+                self.spawn_resolver(node.creds().clone());
+            }
         }
-        self.nodes
-            .get(ip.as_ref()?)
-            .map(|creds| {
-                for cred in creds {
-                    if let Some(CredentialsKind::InterNode(other_name)) = cred.kind() {
-                        if node_name == other_name {
-                            return true;
-                        }
+    }
+
+    fn check_node_request(&self, node_name: &String, ip: Option<SocketAddr>) -> bool {
+        let mut unresolved = None;
+        {
+            if ip.is_none() {
+                return false;
+            }
+            let ip = ip.unwrap().ip();
+            let nodes = self.nodes.read().expect("nodes credentials lock");
+            if let Some(cred) = nodes.get(node_name).map(|guard| guard.creds()) {
+                unresolved = Some(false);
+                if let CredentialsKind::InterNode(other_name) = cred.kind() {
+                    debug_assert!(node_name == other_name);
+                    if cred.ip().iter().find(|cred_ip| cred_ip.ip() == ip).is_some() {
+                        return true;
                     }
+                    
+                    unresolved = Some(true);
                 }
-                false
-            })
+            }
+        }
+        unresolved.map(|unresolved| 
```

Comment:
IMHO, `if let Some(unresolved) = unresolved {` is better from readability perspective